linera-views/src/value_splitting.rs

```diff
@@ -0,0 +1,563 @@
+    K::Error: From<bcs::Error> + From<DatabaseConsistencyError>,
+{
+    /// Creates a new client that deals with big values from one that does not.
+    pub fn new(client: K) -> Self {
+        ValueSplittingKeyValueStoreClient { client }
+    }
+
+    fn read_count_from_value(value: &[u8]) -> Result<u32, K::Error> {
+        if value.len() < 4 {
+            return Err(DatabaseConsistencyError::NoCountAvailable.into());
+        }
+        let mut bytes = value[0..4].to_vec();
+        bytes.reverse();
+        Ok(bcs::from_bytes::<u32>(&bytes)?)
+    }
+
+    fn get_segment_key(key: &[u8], index: u32) -> Result<Vec<u8>, K::Error> {
+        let mut big_key_segment = key.to_vec();
+        let mut bytes = bcs::to_bytes(&index)?;
+        bytes.reverse();
+        big_key_segment.extend(bytes);
+        Ok(big_key_segment)
+    }
+
+    fn read_index_from_key(key: &[u8]) -> Result<u32, K::Error> {
+        let len = key.len();
+        if len < 4 {
+            return Err(DatabaseConsistencyError::TooShortKey.into());
+        }
+        let mut bytes = key[len - 4..len].to_vec();
+        bytes.reverse();
+        Ok(bcs::from_bytes::<u32>(&bytes)?)
+    }
+
+    fn get_initial_count_first_chunk(count: u32, first_chunk: &[u8]) -> Result<Vec<u8>, K::Error> {
+        let mut bytes = bcs::to_bytes(&count)?;
+        bytes.reverse();
+        let mut value_ext = Vec::new();
+        value_ext.extend(bytes);
+        value_ext.extend(first_chunk);
+        Ok(value_ext)
+    }
+}
+
+/// A virtual DB client where data are persisted in memory.
+#[derive(Clone)]
+pub struct TestMemoryClientInternal {
+    client: BasicMemoryClient,
+}
+
+#[async_trait]
+impl KeyValueStoreClient for TestMemoryClientInternal {
+    const MAX_CONNECTIONS: usize = 1;
+    // We set up the MAX_VALUE_SIZE to the artificially low value of 100
+    // purely for testing purposes.
+    const MAX_VALUE_SIZE: usize = 100;
+    type Error = MemoryContextError;
+    type Keys = Vec<Vec<u8>>;
+    type KeyValues = Vec<(Vec<u8>, Vec<u8>)>;
+
+    async fn read_key_bytes(&self, key: &[u8]) -> Result<Option<Vec<u8>>, MemoryContextError> {
+        self.client.read_key_bytes(key).await
+    }
+
+    async fn read_multi_key_bytes(
+        &self,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<Option<Vec<u8>>>, MemoryContextError> {
+        self.client.read_multi_key_bytes(keys).await
+    }
+
+    async fn find_keys_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::Keys, MemoryContextError> {
+        self.client.find_keys_by_prefix(key_prefix).await
+    }
+
+    async fn find_key_values_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::KeyValues, MemoryContextError> {
+        self.client.find_key_values_by_prefix(key_prefix).await
+    }
+
+    async fn write_batch(&self, batch: Batch, base_key: &[u8]) -> Result<(), MemoryContextError> {
+        self.client
+            .write_batch(batch, Self::MAX_VALUE_SIZE, base_key)
```

Comment:
We could just do the size check here, then we wouldn't have to pass in an argument to `write_batch`, and `BasicMemoryClient` could just be `MemoryClient`.

Reply:
Ok.

Reply:
Yes, basically the trait `KeyValueStoreClient` is requiring its implementations
* to declare a `MAX_VALUE_SIZE` limit as a trait constant (so that users of the implementation can adjust)
* to enforce this limit locally to help with debugging.

Avoiding code duplication is important but can be done by defining `Batch::check_size(&self, max_value_size: usize) -> Result<(), SomeError>`.

Reply:
Ok, makes sense.

Reply:
Done.